Internationalization

jonschumacher opened the pull request:
Closes #13

eford commented on src/computational_thinking.jl:
```diff
@@ -14,47 +14,46 @@ export TODO
 
 "Admonition box with reminder that variable name passed is not defined."
-var_not_defined(variable_name) = Markdown.MD(Markdown.Admonition("danger", "Oopsie!", [md"Make sure that you define a variable called **$(Markdown.Code(string(variable_name)))**"]));
-
-"Admonition box with reminder that variable name passed is not defined.  (deprecated)"
-not_defined(variable_name) = var_not_defined(variable_name)
+var_not_defined(variable_name, lang::AbstractLanguage = default_language[]) = Markdown.MD(Markdown.Admonition("danger", var_not_defined_str(lang), [var_not_defined_text_str(lang, variable_name)]));
```
In most other cases, the language parameter is after the other ones.  Is there a reason for these to take lang first?

jonschumacher replied:
Accident. Done!

eford commented on src/i18n.jl:
```diff
@@ -0,0 +1,58 @@
+var_not_defined_text_str(lang::Lang, variable_name) where {Lang <: English} = md"Make sure that you define a variable called **$(Markdown.Code(string(variable_name)))**"
+keep_working_str(lang::Lang) where {Lang <: English} = "Keep working on it!"
+keep_working_text_str(lang::Lang) where {Lang <: English} = md"The answer is not quite right."
+keep_working_update_str(lang::Lang) where {Lang <: English} = md"Make sure to update the cell setting $var."
```
var isn't included in the parameters.  Was that an accident or intentional?

jonschumacher replied:
Accident. Done!

eford commented on src/computational_thinking.jl:
```diff
@@ -66,7 +65,7 @@ function keep_working_if_var_contains_substr(var::Symbol,str::Union{String,Markd
       end
       if occursin(substr,str)
-         keep_working(md"Make sure to update the cell setting $var.")
+         keep_working(keep_working_update_str(lang))
```
Shouldn't we pass var here?

jonschumacher replied:
Done!

eford commented on src/i18n.jl:
```diff
@@ -0,0 +1,58 @@
+wrong_type_text_str(lang::Lang) where {Lang <: English} = md"Check the type of your response."
+wrong_type_text_str(lang::Lang, var, type) where {Lang <: English} = md"The type of $var should be $type"
+func_not_defined_str(lang::Lang) where {Lang <: English} = "Oopsie!"
+func_not_defined_text_str(lang::Lang, func_name) where {Lang <: English} = md"Make sure that you define a function called **$(Markdown.Code(string(func_name)))**"
+var_not_defined_str(lang::Lang) where {Lang <: English} = "Oopsie!"
+var_not_defined_text_str(lang::Lang, variable_name) where {Lang <: English} = md"Make sure that you define a variable called **$(Markdown.Code(string(variable_name)))**"
```
Here are the functions where lang is the first parameter (in case we want to keep lang as the last argument).

jonschumacher replied:
Done!

eford commented on src/i18n.jl:
```diff
@@ -0,0 +1,58 @@
+present_str(lang::Lang) where {Lang <: English} = "present"
+present_mode_str(lang::Lang) where {Lang <: English} = "Present Mode"
+
+default_language = Ref{AbstractLanguage}(EnglishUS())
```
I'm not a frequent user of references in Julia.  Would you mind explaining why you used one here?

jonschumacher replied:
This was mainly based on [this comment](https://github.com/tknopp/RedPitayaDAQServer/commit/cfa713ec2c1248c856603a0d91eb1b353ad885e0#r74307377). When looking deeper into this, I could not find details on this. I am also not sure how much penalty we get by having this variable type Any. With Julia 1.8, we could also directly annotate the type to avoid the instability but this would break compatability with older Julia versions.

eford commented on src/PlutoTeachingTools.jl:
```diff
@@ -2,10 +2,12 @@ module PlutoTeachingTools
 using PlutoLinks: @ingredients
 export @ingredients
 
+include("i18n.jl")
```
I'd like to add a comment like
# provides framework for multiple language support
for people who don't recognize i18n.

jonschumacher replied:
Done!

eford commented on src/i18n.jl:
```diff
@@ -0,0 +1,58 @@
+
+
+
+abstract type English  <: AbstractLanguage end
```
I notice that you have the English *_str functions in the same module and file.  
I was originally thinking that it might be nice to have a separate module for each language, so that there's more parallel treatment of different languages.  Is there a disadvantage to that?

eford replied:
Even if we end up putting all the *_str functions in one module, I'd suggest that we organize them into separate files.  That way it'll be easier when someone submits a pull request to see that they just added `include("spanish.jl")` to this file and a bunch of stuff in spanish.jl.

jonschumacher replied:
Done!

eford commented:
Thanks.  I skimmed it and think this looks very promising.   I asked some questions where I wasn't sure about the pros/cons of the different ways we could implement something.   

In order to minimize the risk of breaking changes in the future, I think it would be good to have at least one other language implemented before we merge the changes.  That way we'll be more likely to notice any complications that we should address now.

jonschumacher commented:
Thanks for the quick review. The draft was mainly meant to inform you where I am doing stuff, but I think I now addressed all the issues. The tests are running through and after setting the language, I get the correct translations. From my perspective we can merge. What do you think?

eford commented:
Thanks!  I tried using it with one of the notebooks for my current class.  I got several error messages like
```UndefVarError: lang not defined
correct()@computational_thinking.jl:137
top-level scope@[Local: 1](http://localhost:1234/edit?id=efd44382-2b35-11ed-397b-4188d5f532fa#)[inlined]
```
So far, all the problems I noticed could be traced to a function where a (currently) required argument has a default value.  In the PR, the default value is set by a function call that depends on lang, but lang is only set in a subsequent argument.  For example
 
```julia
correct(text=rand(yays(lang)), lang::AbstractLanguage = default_language[]) = Markdown.MD(Markdown.Admonition("correct", correct_str(lang), [text]));
```

The other cases I noticed were `protip`, `still_missing`, `still_nothing` and `keep_working`.
I see two potential solutions:
1.  Make any optional parameter a named parameter
```julia
correct(; lang::AbstractLanguage = default_language[], text=rand(yays(lang)) ) = Markdown.MD(Markdown.Admonition("correct", correct_str(lang), [text]));
```
Pros:
- Seems workable and clean
- Would make it easy to generalize further if we wanted to allow people to override the hint_str, tip_str, etc. manually.  (But If people override the default hint_str, tip_str, etc. it won't be easy to swap out languages.  So maybe its better to not to generalize functions like `hint` and `tip`.)

Cons:
- This would be a breaking change to the API.  (Ok with me in this case.)
- People who manually set lang would have to write `lang=` over and over again.   (Maybe not an issue if we think people will just use `set_language!` and not specify it manually.)

Open Question:
- Would we make lang a named argument for every function for consistency, even though most functions don't need it that way?  If not, people would need to know which functions have named parameters.  

2.  For the functions that currently have arguments with default values (other than lang), make two versions (one with all named arguments, one with unnamed arguments), so as to preserve backward compatibility and prevent people from having to use named arguments.
```julia
correct(; lang::AbstractLanguage = default_language[], text=rand(yays(lang)) ) = Markdown.MD(Markdown.Admonition("correct", correct_str(lang), [text]));
correct(text, lang::AbstractLanguage = default_language[]) = correct(;lang=lang, text=text)
```
Pros:
- Makes changes non-breaking to API
- Could do this just for the ~5 functions that currently need it, and still have option of doing something similar for other functions down the road.

Cons:
- A few more lines of code to maintain

Currently, I'm leaning towards option #2.  
Any thoughts?

jonschumacher commented:
I now opted for opion 2 and implemented it. Those few more lines should not be hard to maintain. I now also added a separation between formal and colloquial German.

eford reviewed (approved):
Thanks
